### PR TITLE
Persist admin sessions so a restart does not force another login

### PR DESCRIPTION
**Pull request by moshouhot**
## Summary

The management console keeps its session table in memory only, so restarting the
gateway invalidates the `cb_admin_session` cookie and the WebUI asks for the API key
again. This is easy to hit during an update or a plain restart.

This persists the session table to `admin-sessions.json` under the managed auth
directory (the same directory as `control.sqlite3` / `logs.sqlite3`, which is already
git-ignored) and restores it on startup, so a restart no longer forces another login.

## Why expiry moves off the monotonic clock

`time.monotonic` restarts with the machine and cannot express an absolute deadline
across restarts, so a persisted monotonic timestamp would be meaningless. Session
expiry now uses the wall clock. Login throttling still uses `time.monotonic`, which
is the appropriate clock for a sliding window and is unaffected by wall-clock jumps.

## Revocation is unchanged

Persisting a login must not weaken the existing guarantees. The design keeps
revocation as the single mechanism — **clearing the stored snapshot**:

- The file carries a keyed fingerprint naming the key epoch it belongs to.
- A snapshot whose fingerprint does not match the current key is **revoked**, not
  adopted. That covers key rotation, a disabled key, and a key that is set back to a
  previous value, across restarts and in-process.
- Logout and expiry clear the entry, and the file is rewritten.
- A snapshot that fails validation is **removed** rather than ignored, so it cannot
  be adopted by a later start.

The fingerprint is an epoch label, not an integrity MAC — revocation is performed by
clearing the file, and the header identity remains ephemeral as before.

## Hardening of the restored snapshot

The snapshot is untrusted input read from disk, so it is:

- read with a bounded read, via `O_NOFOLLOW` with a regular-file check, so a symlink
  is not followed and an oversized file is not loaded;
- parsed as strict JSON (duplicate keys rejected, `Infinity` / `NaN` rejected);
- validated field by field: exact top-level and per-entry key sets, integer version,
  bounded url-safe tokens, and a finite in-range deadline with `bool` rejected;
- size bounded in both entry count and bytes.

Writes go through a mode-0600 temporary file and an atomic `os.replace`. If the write
cannot complete, the previous snapshot is revoked as the fallback, so a stale
on-disk table cannot outlive a revocation. Note that `chmod` does not establish
owner-only ACLs on Windows; the mode is best effort there.

When no `session_path` is configured the previous in-memory behaviour is unchanged.

## Tests

`tests/test_admin_boundaries.py` gains a `PersistedSessionTests` class covering:
survival across a restart, the restored CSRF token and deadline, expiry under an
injected wall clock, key rotation, a disabled key both at startup and in-process,
logout, a failing write, symlinks, and a table of untrusted snapshots
(`Infinity` / `NaN` expiry, duplicate and extra fields, boolean version, non-ASCII
fingerprint, oversized sid / csrf / file).

`python -m pytest tests/test_admin_boundaries.py tests/test_admin_api.py` → 62 passed.
Full suite on Windows shows the same 20 pre-existing failures before and after this
change (symlink / fsync / Docker / permission tests); verified by diffing the failure
list against pristine `main`. No new failures.

## Summary by Sourcery

Persist admin sessions across restarts while preserving durable revocation and failing safely when session storage is unavailable.

New Features:
- Persist administrator sessions to disk and restore them across gateway restarts without requiring another API-key login.

Bug Fixes:
- Ensure session expiry remains correct across restarts and that key rotation, disabled keys, logout, expiry, and invalid snapshots revoke sessions durably.
- Fail closed when a superseded session snapshot cannot be revoked, preventing stale sessions from being resurrected.

Enhancements:
- Harden persisted session handling with bounded, strict validation and atomic owner-protected snapshot updates.
- Expose session-storage health in the administration API and surface failed revocations in the WebUI.

Tests:
- Add coverage for restart persistence, expiry, revocation, storage failures, symlink handling, malformed snapshots, bounds, and UI storage-status behavior.

### CODE REVIEW

File: app/admin_auth.py [hunk 1]
```diff
@@ -64,24 +114,161 @@ def same_origin(request, allowed=()):
...
+    def _token(value):
+        """Accept only bounded url-safe tokens, rejecting bools and non-strings."""
+        return value if isinstance(value, str) and _TOKEN.match(value) else None
+
+    @staticmethod
+    def _deadline(value):
+        """Accept only finite, in-range numeric deadlines; bool is not a deadline."""
+        if type(value) not in (int, float):
+            return None
+        number = float(value)
```
[the remark this answers is not in the document]
Comment (moshouhot) [reply app/admin_auth.py]: Fixed in a4da831. `_deadline` now returns `None` on `OverflowError`, so an out-of-float-range integer is treated as an invalid deadline and the snapshot is revoked rather than raising.

You were right that the exception escaped the revoke path too: `_key()` assigns `_configured_key` before `_restore` runs, so the failure was worse than a single crash — the next call in the same process skipped restoration and returned `True` with an empty table, while the file stayed on disk, contradicting the "a snapshot that fails validation is removed" guarantee. Regression coverage is in `test_untrusted_snapshots_are_revoked_rather_than_adopted` (`expiry beyond float`, `negative expiry beyond float`), which asserts the file is gone afterwards.

File: app/admin_auth.py [hunk 2]
```diff
@@ -64,24 +114,161 @@ def same_origin(request, allowed=()):
...
+                if not stat.S_ISREG(metadata.st_mode) or metadata.st_size > _SESSION_FILE_BYTES:
+                    return False
+                raw = stream.read(_SESSION_FILE_BYTES + 1)
+        except OSError:
+            return False
+        if len(raw) > _SESSION_FILE_BYTES:
+            return False
+        try:
+            document = _strict_json(raw)
+        except (ValueError, UnicodeDecodeError):
```
[the remark this answers is not in the document]
Comment (moshouhot) [reply app/admin_auth.py]: Fixed in a4da831. `RecursionError` is now caught alongside `ValueError` / `UnicodeDecodeError`, so deeply nested input inside the size bound is malformed input, not a startup failure.

Reproduced as described: 100 KB of `[` x 50 000 (well inside the 256 KiB limit) raised out of `json.loads`, and because the exception bypassed the `_revoke()` branch the file survived for the next start. `test_untrusted_snapshots_are_revoked_rather_than_adopted` now carries a `deeply nested` case and asserts the snapshot is removed.

File: app/admin_auth.py [hunk 3]
```diff
@@ -64,24 +114,161 @@ def same_origin(request, allowed=()):
...
+        """Load persisted sessions for the current key epoch.
+
+        Returns False when a snapshot exists that cannot be trusted or validated,
+        so the caller revokes it instead of leaving it available to a later start.
+        """
+        if self._path is None:
+            return True
+        try:
+            fd = os.open(self._path, os.O_RDONLY | getattr(os, "O_NOFOLLOW", 0) | getattr(os, "O_NONBLOCK", 0))
+        except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (moshouhot) [reply app/admin_auth.py:216]: Fixed in a4da831, and you are right that this was a real gap rather than a theoretical one — confirmed on Windows: with a valid snapshot at the target, `_restore` adopted the session and `session()` returned it.

`_is_direct_file` now rejects a path that `os.lstat` reports as a link and, as a fallback, requires the opened file's `(st_dev, st_ino)` to match the path's, so the no-symlink guarantee holds where `O_NOFOLLOW` does not exist. The existing test only pointed at a plain file, which could not distinguish "rejected" from "parsed and discarded", so it now points at a valid snapshot and asserts both that the session was not adopted and that the target is untouched.

One residual worth stating plainly, since the PR body is explicit that the fingerprint is an epoch label and not an integrity MAC: a local actor with write access to the auth directory can still replace the file with a crafted snapshot carrying a copied fingerprint, because reading a valid snapshot reveals the fingerprint without the API key. Rejecting links closes the path that did not require write access to the snapshot itself, but the directory remains the trust boundary — the same level as `control.sqlite3` and `logs.sqlite3`. I have left that as documented behaviour rather than adding a MAC; say the word if you would rather key the fingerprint over the session contents.

File: app/admin_auth.py [hunk 4]
```diff
@@ -64,24 +114,161 @@ def same_origin(request, allowed=()):
...
+                return False
+            token = self._token(sid)
+            csrf_token = self._token(item["csrf_token"])
+            expires = self._deadline(item["expires"])
+            if token is None or csrf_token is None or expires is None:
+                return False
+            if expires > now:
+                restored[token] = {"csrf_token": csrf_token, "expires": expires}
+        self.sessions.update(restored)
+        return True
```
[the remark this answers is not in the document]
Comment (moshouhot) [reply app/admin_auth.py:271]: Fixed in a4da831. When restoration drops any entry, `_restore` calls `_persist()` so the expired records are gone from disk, not only from `self.sessions`.

Reproduced: after a restart past the deadline the in-memory table was empty while the file still held the record, and a further restart with the clock moved earlier revived it. `test_expired_records_are_dropped_from_the_snapshot` covers that sequence, and `test_expiry_uses_the_injected_wall_clock_across_instances` now asserts the record is absent from the rewritten file.

I kept the guard as `len(restored) != len(entries)` rather than comparing full contents — a snapshot can only lose entries at that point, since every surviving entry is re-emitted verbatim.

**Comment by maiphucgiang** [comment]
@codex review

File: app/admin_auth.py [hunk 5]
```diff
@@ -64,24 +114,186 @@ def same_origin(request, allowed=()):
...
+            restoring = self._configured_key is None
             self._configured_key = key
+            self.sessions.clear()
             # Restoring a previous key must not restore that epoch's OAuth owner.
             self._identity = secrets.token_urlsafe(32)
+            if restoring:
+                # Adopt only a snapshot that matches the current epoch; anything else is
+                # revoked, so switching back to an old key cannot resurrect its sessions.
+                if not self._restore(key):
+                    self._revoke()
```
Comment (maiphucgiang) [inline app/admin_auth.py]: [P1] Revoke obsolete snapshots at startup, not on the first admin request

`_key()` is lazy: `AdminAuth.__init__()` and `install_admin()` do not call it, and inference requests bypass `AdminMiddleware`. Consequently, log in with key A, restart with key B (or an empty key), make only inference requests, and restart again with A before the cookie expires: the original cookie is restored and accesses a protected admin endpoint with HTTP 200. I reproduced both intermediate-key variants using the real admin and inference middleware; main returns 401. The intermediate process never clears A's snapshot, so the deterministic fingerprint cannot distinguish the reused key from the old epoch.

Please reconcile/revoke the persisted snapshot during startup after the effective configuration is resolved, including when management is disabled. Add restart tests for A → B → A and A → empty → A that deliberately make no admin request during the intermediate run.

File: app/admin_auth.py [hunk 6]
```diff
@@ -140,15 +353,17 @@ def login(self, request, key):
...
+            while len(self.sessions) > MAX_PERSISTED_SESSIONS:
                 self.sessions.popitem(last=False)
+            self._persist()
             return (sid, dict(item)), 200
 
     def logout(self, request):
         with self.lock:
-            self.sessions.pop(request.cookies.get(COOKIE_NAME), None)
+            if self.sessions.pop(request.cookies.get(COOKIE_NAME), None) is not None:
+                self._persist()
```
Comment (maiphucgiang) [inline app/admin_auth.py]: [P1] Do not report successful logout when durable revocation fails

The result of `_persist()` is discarded. If the session directory is not writable, both the replacement write and the fallback `_revoke()`/`unlink()` fail, leaving the valid session on disk. I reproduced this with a real directory permission change under a non-root user: `DELETE /admin/session` still returned 200 with `authenticated: false`; after restoring permissions and restarting, replaying the saved cookie returned 200 from a protected admin endpoint. The same sequence returns 401 on main.

Please propagate durable-revocation failure to the logout endpoint instead of acknowledging completion, keep the failed operation retryable, and expose the failure in the audit/degraded state. Add coverage where both writing and unlinking fail; the existing write-failure test allows unlink to succeed and therefore misses this case.

File: tests/test_admin_boundaries.py [hunk 7]
```diff
@@ -79,6 +79,227 @@ def test_key_comparison_remains_constant_time(self):
...
+            first.logout(request(cookie=sid))
+        self.assert_revoked(sid)
+
+    def test_key_rotation_survives_a_write_failure(self):
+        sid, _ = self.login(AdminAuth(self.config))
+        with patch("app.admin_auth.tempfile.mkstemp", side_effect=OSError("disk full")):
+            AdminAuth({"api_key": "rotated-synthetic-key", "session_path": self.path}).enabled()
+        self.assert_revoked(sid)
+
+    def test_symlinked_snapshot_is_never_followed(self):
```
Comment (maiphucgiang) [inline tests/test_admin_boundaries.py]: [P3] Give the strengthened symlink test a distinct name

`PersistedSessionTests` already defines `test_symlinked_snapshot_is_never_followed` at line 179. This second definition overwrites the first before test collection, so the strengthened case pointing at a valid session snapshot never runs; only this weaker sentinel-file case runs. Please give the tests distinct names (or combine their assertions), and verify that the valid-snapshot case is collected.

**Review by maiphucgiang — CHANGES_REQUESTED** [review]
Please address the two session-revocation regressions below before merging, and restore the overwritten symlink test. I reviewed a4da831 and ran the focused admin tests (63 passed). Additional isolated reproductions against this head and main@426e8a5 confirmed that both revocation scenarios return HTTP 200 for an old cookie on this branch, versus HTTP 401 on main. Only synthetic credentials and temporary storage were used; no upstream services were contacted.

File: app/admin_auth.py (hunk 5 above)
Comment (moshouhot) [reply app/admin_auth.py]: Fixed in 79386ba. `install_admin` now calls `AdminAuth.reconcile()`, which resolves the key epoch during startup instead of on the first `/admin` request.

You were right about the mechanism: `_key()` is lazy, `AdminAuth.__init__()` and `install_admin()` never called it, and `AdminMiddleware` only covers `/admin/`, so an inference-only process never touched the file and the next start adopted it. `reconcile()` is called before the middleware is installed, so it also runs when management is disabled and the key is empty.

Verified against your scenario through the real ASGI app: A → B (inference only) → A now returns 401 from `/admin/settings` for the old cookie, where it previously returned 200. Both restart tests you asked for are in `test_startup_reconciles_without_any_admin_request` (A → B → A and A → empty → A, each asserting no admin request is made during the intermediate run), plus `test_startup_reconcile_happens_for_the_installed_admin` to keep the wiring from regressing.

Your point about the deterministic fingerprint is exactly why this could not be fixed by inspecting the file: with A and B both valid epochs, nothing on disk distinguishes "reused key" from "same epoch". The fix has to be that the intermediate process actually runs its reconciliation.

File: app/admin_auth.py (hunk 6 above)
Comment (moshouhot) [reply app/admin_auth.py]: Fixed in 79386ba. `logout()` now returns whether the revocation reached disk, and `session_delete` answers 503 rather than `authenticated: false` when it did not.

Two changes beyond propagating the flag, because propagating alone would have been misleading:

- The in-memory entry is kept when the snapshot cannot be rewritten, so the session is still live and a retry can finish the revocation. Previously the entry was dropped from `self.sessions` regardless, so a retry would have found nothing to revoke and the cookie would have stayed valid on disk until expiry.
- The failure is recorded as session-store state and surfaced in `/admin/settings` as `session` (`{path, degraded, last_error}`), mirroring the existing `audit` state, and the settings page shows a warning panel. That covers "expose the failure in the degraded state" without inventing a new mechanism.

Coverage where both paths fail: `test_logout_reports_a_revocation_that_could_not_be_persisted` patches `mkstemp` **and** `os.unlink`, asserts `logout()` is False, that the session is still usable, that a retry after storage recovers succeeds, and that the snapshot is gone; `test_logout_reports_when_the_session_could_not_be_persistently_revoked` asserts 503 over HTTP and that the cookie still works before the retry. You were right that the old write-failure test allowed unlink to succeed and therefore never reached this case.

Reproduced end-to-end with a read-only directory: `DELETE /admin/session` now returns 503, and `/admin/settings` reports `session.degraded: true`.

File: tests/test_admin_boundaries.py (hunk 7 above)
Comment (moshouhot) [reply tests/test_admin_boundaries.py]: Fixed in 79386ba. You are right, and this one is on me — I added the strengthened case without noticing the original definition later in the file, so Python's later binding silently replaced it and only the sentinel-file case was collected. The regression test I described in my earlier reply never ran.

The two cases are now one test with a subtest per target: a link to a plain file, and a link to a valid session snapshot, both asserting the target is neither adopted (`dict(auth.sessions) == {}`) nor modified. I checked collection rather than assuming: `test_symlinked_snapshot_is_never_followed` appears once, and the valid-snapshot assertions run.

I also added a duplicate-method scan over the test file while fixing this; it reports no other shadowed definitions.

File: app/admin_auth.py (hunk 5 above)
Comment (maiphucgiang) [reply app/admin_auth.py]: > Fixed in [79386ba](https://github.com/maiphucgiang/codebuddy2api/commit/79386ba6f6979a94627b6a644d4e6fdd63dcd2a1). `install_admin` now calls `AdminAuth.reconcile()`, which resolves the key epoch during startup instead of on the first `/admin` request.
> 
> You were right about the mechanism: `_key()` is lazy, `AdminAuth.__init__()` and `install_admin()` never called it, and `AdminMiddleware` only covers `/admin/`, so an inference-only process never touched the file and the next start adopted it. `reconcile()` is called before the middleware is installed, so it also runs when management is disabled and the key is empty.
> 
> Verified against your scenario through the real ASGI app: A → B (inference only) → A now returns 401 from `/admin/settings` for the old cookie, where it previously returned 200. Both restart tests you asked for are in `test_startup_reconciles_without_any_admin_request` (A → B → A and A → empty → A, each asserting no admin request is made during the intermediate run), plus `test_startup_reconcile_happens_for_the_installed_admin` to keep the wiring from regressing.
> 
> Your point about the deterministic fingerprint is exactly why this could not be fixed by inspecting the file: with A and B both valid epochs, nothing on disk distinguishes "reused key" from "same epoch". The fix has to be that the intermediate process actually runs its reconciliation.


Thanks — I verified the writable-storage rotation cases, failed-logout retry, and duplicate-test fix on 79386ba; those now pass. One P1 failure path in this thread remains.

**[P1] Startup still succeeds when revoking the previous epoch fails**

[`reconcile()` at lines 331–340](https://github.com/maiphucgiang/codebuddy2api/blob/79386ba6f6979a94627b6a644d4e6fdd63dcd2a1/app/admin_auth.py#L331-L340) calls `_key()`, but `_key()` still discards the failed `_revoke()` result at line 317. `install_admin()` therefore completes successfully even though the previous epoch's snapshot survives.

I reproduced this with a real read-only bind mount, not mocked I/O errors:

1. Start with a valid persisted session for key A, then make only `admin-sessions.json` read-only. The containing directory and sibling SQLite database remain writable.
2. Install the real admin/inference middleware with key B (also tested with an empty key), and make only inference requests. Startup succeeds and inference returns 200; session storage merely reports `degraded: true`.
3. Start again with A and replay the old cookie against a protected admin endpoint: **200**, versus **401** on the non-persistent baseline.

So the original issue is fixed when revocation succeeds, but a storage failure still permits resurrection. A settings-page warning cannot prevent the next process from adopting the surviving file, and that page is unavailable when the intermediate key is empty.

Please propagate durable-revocation failure through `_key()` → `reconcile()` → `install_admin()`. If an obsolete snapshot cannot be removed or replaced, startup must explicitly fail rather than activate the new key epoch, unless an independent durable revocation mechanism guarantees the old snapshot cannot return. Add failure-path tests for A → B → A and A → empty → A with an undeletable snapshot and writable SQLite, while retaining the successful-reconciliation tests.

Comment (moshouhot) [reply app/admin_auth.py]: Fixed in e0bd154. You are right, and thank you for the follow-up: my 79386ba only recorded the failure as degraded state, so the rotation still completed and the snapshot still survived — the file was left for a later start to adopt. Reproduced exactly as you describe, with a read-only `admin-sessions.json` and the directory and sibling SQLite stores left writable: `install_admin` for the new key returned normally, and a later start under the old key accepted the cookie.

`_key()` now raises `SessionStoreError` when neither the rewrite nor the `_revoke()` fallback made the revocation durable, and rolls `_configured_key` back so the epoch stays unactivated. That rollback is deliberate: without it the next call would see the new key already configured, skip the revocation branch, and report success — the same "recorded as done" bug one level down. Because the epoch is not activated, a retry after the permissions are fixed does the revocation for real instead of finding nothing to do.

Propagation, since a stored flag was exactly the thing that failed:

- `converter.main` catches it and aborts startup with a readable reason (exit 2) rather than serving traffic — verified in a subprocess: it refuses with the undeletable snapshot and still starts and serves `/health` when the snapshot is valid under a matching key.
- `AdminMiddleware` catches it and answers 503, so if an epoch change degrades after startup, management traffic fails closed rather than open. Confirmed that repeated requests all return 503, not just the first.

The tests use real permissions rather than mocked I/O, per your "distinct storage-failure mode" point: a read-only snapshot file, plus a read-only directory where POSIX allows the unlink (Windows blocks the unlink on the read-only file alone), with the sibling stores writable. `test_startup_fails_when_a_superseded_snapshot_cannot_be_revoked` covers rotation and disabled-key cases, the failed retry, and recovery after permissions are restored; `test_startup_failure_reaches_the_installed_admin` covers propagation, and `test_a_superseded_snapshot_that_cannot_be_revoked_fails_closed` covers the 503. I checked the startup test against the previous `_key()` behaviour and it fails there, so it does pin the fix.

The old write-failure tests still pass unchanged, so the mocked path and the real-permission path agree.

Verification: `tests/test_admin_boundaries.py`, `tests/test_admin_api.py`, `tests/test_runtime_endpoints.py`, `tests/test_inference_admission.py` — 127 passed; `pnpm test` — 132 passed; the full Python suite matches its baseline apart from the pre-existing flaky trial-ledger concurrency test, which also fails on `main`. I also re-ran the duplicate-method scan, since shadowed tests were the P3 finding: none.

One judgement call worth flagging: I made this fatal at startup rather than degrading, because the alternative — serving with the epoch switch recorded as complete — is what created the resurrection path. That means a permissions problem in the auth directory now stops the gateway from starting. If you would rather it start read-only and simply refuse management traffic, that is a small change and I am happy to switch it.

Comment (moshouhot) [reply app/admin_auth.py]: Two follow-ups since that reply, both on the same head:

**3a8a3e5 — the fixture was not portable, and would have failed on your CI.** Your "distinct storage-failure mode" point made me re-check my own helper, and it had the same class of bug I was fixing: it verified the lock by unlinking the snapshot itself. On Windows the read-only file mode blocks that unlink, so the probe doubled as the check; on POSIX the file mode does not affect removal, so it deleted the fixture and the test asserted nothing while still passing. It also re-created the probe *after* chmod'ing it read-only, which raises on POSIX. The check now runs against a throwaway probe file, the directory is locked only after that check shows the file mode was not enough, and platforms where removal cannot be blocked skip rather than pass silently. Verified against simulated POSIX semantics (unlink gated on directory write permission) and against a platform where every unlink succeeds: the POSIX path locks the directory, keeps the snapshot undeletable, refuses the reconcile and restores the mode; the unblockable path skips. I cannot run Linux here, so that verification is by simulation rather than on your runner — worth watching the CI result for this one.

**a5edf2a — merged `main`.** Upstream added the usage-snapshot publish before `runtime_management.install`, which collides with the `SessionStoreError` guard in `converter.py`; the PR had gone `CONFLICTING`. Both changes are kept: publish cached usage, then install under the guard. It is `MERGEABLE` again, and `tests/test_usage_snapshots.py` and `tests/test_credential_cooldowns.py` from that PR pass alongside the admin tests (198 passed, 1 skipped).

Full-suite note: 937 passed, and the failure set is identical to the pre-change baseline apart from the pre-existing flaky `test_trial_rewards` concurrency test. One further flake I saw once and chased down rather than dismissing: `test_observability.py::test_first_effective_output_not_role_or_usage` asserts `first_token_ms >= 20` after `asyncio.sleep(0.02)`, and on Windows that sleep quantises to one timer tick — I measured `time.monotonic()` steps of exactly 16.0 ms and 4/400 samples below 20 ms, and the failure I captured was `15.999999988707714`. `app/observability.py` and `tests/test_observability.py` are byte-identical to `main` on this branch, so it is not from this change; flagging it only because it is a real Windows-only flake in a test you run.

Comment (moshouhot) [reply app/admin_auth.py]: Following up on the one caveat I left open: CI has now run on the merged head and the Linux `test` job is green, so the POSIX path is verified on a real runner rather than only by simulation. Specifically, the three tests that exercise the storage-failure path all report `ok` on Linux and none of the admin tests skipped — `test_startup_fails_when_a_superseded_snapshot_cannot_be_revoked`, `test_startup_failure_reaches_the_installed_admin`, and `test_logout_reports_a_revocation_that_could_not_be_persisted`. The directory-mode branch in the fixture is therefore exercised for real, which was the part I could not check locally.

All checks on `a5edf2a` pass: `test`, `web`, `image`, CodeQL (all three languages) and Sourcery. The PR is `MERGEABLE`; `CHANGES_REQUESTED` is still standing on your side, so it is yours to clear when you are satisfied.
